src/logs.ts

```diff
@@ -0,0 +1,345 @@
+   * Maps log listener id to callback.
+   */
+  private _logCallbacks: Map<number, LogCallback>;
+
+  /**
+   * The subscription id from the connection onLogs subscription.
+   */
+  private _onLogsSubscriptionId: number | undefined;
+
+  constructor(programId: PublicKey, connection: Connection) {
+    this._programId = programId;
+    this._connection = connection;
+    this._eventCallbacks = new Map();
+    this._logCallbacks = new Map();
+    this._listenerIdCount = 0;
+  }
+
+  /**
+   *
+   * @param abi
+   * @param callback
+   * @returns
+   */
+  public addEventListener(
+    abi: ethers.utils.Interface,
+    callback: EventCallback
+  ): number {
+    let listener = this.getNewListenerId();
+
+    // Store the callback into the events listeners map.
+    this._eventCallbacks.set(listener, [abi, callback]);
+
+    // Create the subscription singleton, if needed.
+    if (this._onLogsSubscriptionId !== undefined) {
+      return listener;
+    }
+
+    this.processLogs();
+
+    return listener;
+  }
+
+  /**
+   *
+   * @param listener
+   */
+  public async removeEventListener(listener: number): Promise<void> {
+    // Get the callback.
+    const callback = this._eventCallbacks.get(listener);
+    if (!callback) {
+      throw new Error(`Event listener ${listener} doesn't exist!`);
+    }
+
+    this._eventCallbacks.delete(listener);
+
+    await this.stopProcessingLogs();
+  }
+
+  /**
+   *
+   * @param callback
+   * @returns
+   */
+  public addLogListener(callback: LogCallback): number {
+    let listener = this.getNewListenerId();
+
+    // Store the callback into the log listeners map.
+    this._logCallbacks.set(listener, callback);
+
+    // Create the subscription singleton, if needed.
+    if (this._onLogsSubscriptionId !== undefined) {
+      return listener;
+    }
+
+    this.processLogs();
+
+    return listener;
+  }
+
+  /**
+   *
+   * @param listener
+   */
+  public async removeLogListener(listener: number): Promise<void> {
+    // Get the callback.
+    const callback = this._logCallbacks.get(listener);
+    if (!callback) {
+      throw new Error(`Log listener ${listener} doesn't exist!`);
+    }
+
+    this._logCallbacks.delete(listener);
+
+    await this.stopProcessingLogs();
+  }
+
+  private getNewListenerId() {
+    this._listenerIdCount += 1;
+    return this._listenerIdCount;
+  }
+
+  private processLogs() {
+    this._onLogsSubscriptionId = this._connection.onLogs(
+      this._programId,
+      (logs, ctx) => {
+        if (logs.err) {
+          return;
+        }
+        for (const log of logs.logs) {
+          const eventData = parseLogTopic(log);
+          const msg = parseLogLog(log);
+
+          if (eventData) {
+            for (const [abi, callback] of this._eventCallbacks.values()) {
+              let event: ethers.utils.LogDescription | null = null;
+              try {
+                event = abi.parseLog(eventData);
+              } catch (e) {
+                // console.log(e);
```

Comment:
What can cause a silent error here?

Reply:
If there is a try catch then the catch should be handled. We shouldn't just swallow errors.

Reply:
I guess we just log the error for now or? Since it's totally expected for a discovered emitted event log not to be parseable by any of the registered listeners abis...